xwiki-platform-core/xwiki-platform-mentions/xwiki-platform-mentions-default/src/main/java/org/xwiki/mentions/internal/async/MapBasedLinkedBlockingQueue.java

```diff
@@ -0,0 +1,364 @@
+     */
+    private Long computeHighestKey()
+    {
+        return this.internalMap.keySet()
+            .stream()
+            // Sort in descending order to get the highest element first
+            .sorted((a, b) ->
+            {
+                if (a.equals(b)) {
+                    return 0;
+                } else if (a < b) {
+                    return 1;
+                } else {
+                    return -1;
+                }
+            })
+            .limit(1).findFirst().orElse(0L);
+    }
+
+    private class DefaultComparator implements Comparator<Map.Entry<Long, T>>
+    {
+        @Override
+        public int compare(Map.Entry<Long, T> a, Map.Entry<Long, T> b)
+        {
+            Long keyA = a.getKey();
+            Long keyB = b.getKey();
+
+            if (keyA.equals(keyB)) {
+                return 0;
+            } else if (keyA < keyB) {
+                return -1;
+            } else {
+                return 1;
+            }
+        }
+    }
+
+    /**
+     * Populate the internal queue representation with the information coming from the Map.
+     * The idea is to be able to reuse entirely the original {@link LinkedBlockingQueue} method without really needing
+     * the map.
+     */
+    private void populateUnderlyingQueue()
+    {
+        this.internalMap.entrySet()
+            .stream()
+            .sorted(new DefaultComparator())
+            .forEach((entry) ->
+            {
+                try {
+                    this.internalQueue.put(Pair.of(entry));
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            });
+    }
+
+    /**
+     * Put an element in the internal map and with the incremented highest key.
+     *
+     * @param t an element to add in the internal map.
+     * @return the actual key used to put the info in the map.
+     */
+    private Long internalPut(T t)
+    {
+        Long key = this.highestKey.incrementAndGet();
+        this.internalMap.put(key, t);
+        return key;
+    }
+
+    /**
+     * Remove an element from the internal map and decrement the highest key only if there's no more element
+     * associated to this key.
+     *
+     * @param pair the element to remove, should be of type Pair.
+     */
+    private void internalRemove(Object pair)
+    {
+        if (pair != null && pair instanceof Pair) {
+            Pair<Long, T> castedPair = (Pair) pair;
+            this.internalMap.remove(castedPair.getKey());
+        }
+    }
+
+    /**
+     * Clear the internal map and reset the highest key to 0.
+     */
+    private void clearInternalMap()
+    {
+        this.internalMap.clear();
+        this.highestKey.set(0);
+    }
+
+    @Override
+    public void put(T t) throws InterruptedException
+    {
+        Long key = this.internalPut(t);
+        this.internalQueue.put(Pair.of(key, t));
```

Comment:
I would put in the key before the map so that it's available faster to whoever is waiting for new entries.

Reply:
It might also lead to some inconsistency: you would retrieve faster the new entry to whoever is waiting, but since you would use the same blocking queue you'd need to remove the key from the map. So you could have a race condition leading to having the entry in the map while it has already been consumed.

Reply:
Yes you are right, it's safer like this.